sdk/core/core/src/az_http_policy_retry.c

```diff
@@ -0,0 +1,228 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "az_http_policy_retry_private.h"
+#include "az_http_private.h"
+#include <az_config.h>
+#include <az_config_internal.h>
+#include <az_http_internal.h>
+#include <az_log_internal.h>
+#include <az_platform_internal.h>
+#include <az_retry_internal.h>
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include <_az_cfg.h>
+
+static az_http_status_code const _default_status_codes[] = {
+  AZ_HTTP_STATUS_CODE_REQUEST_TIMEOUT,
+  AZ_HTTP_STATUS_CODE_TOO_MANY_REQUESTS,
+  AZ_HTTP_STATUS_CODE_INTERNAL_SERVER_ERROR,
+  AZ_HTTP_STATUS_CODE_BAD_GATEWAY,
+  AZ_HTTP_STATUS_CODE_SERVICE_UNAVAILABLE,
+  AZ_HTTP_STATUS_CODE_GATEWAY_TIMEOUT,
+  AZ_HTTP_STATUS_CODE_NONE,
+};
+
+AZ_NODISCARD az_http_policy_retry_options az_http_policy_retry_options_default()
+{
+  return (az_http_policy_retry_options){
+    .max_retries = 4,
+    .retry_delay_msec = 4 * _az_TIME_MILLISECONDS_PER_SECOND, // 4 seconds
+    .max_retry_delay_msec
+    = 2 * _az_TIME_SECONDS_PER_MINUTE * _az_TIME_MILLISECONDS_PER_SECOND, // 2 minutes
+    .status_codes = _default_status_codes,
+  };
+}
+
+AZ_NODISCARD bool az_http_policy_retry_options_validate(
+    az_http_policy_retry_options const* retry_options)
+{
+  if (retry_options == NULL || retry_options->max_retries < 0
+      || retry_options->status_codes == NULL)
+  {
+    return false;
+  }
+
+  if (retry_options->retry_delay_msec < 0 || retry_options->max_retry_delay_msec < 0)
```

Comment:
How is the test coverage for this new code/policy?

Reply:
https://github.com/Azure/azure-sdk-for-c/issues/416

Reply:
I consider this feedback as 'done' in the scope of this PR.
Please revert if you disagree.

As I said during the standup, I'll try to implement az_platform_test soon, and that will be a separate PR.